### PR TITLE
Added support for logging in using shared web credentials

### PR DESCRIPTION
**Pull request by clausjoergensen**
Follow-up to #314 where we fetch the credentials from the shared web credentials, allowing the user to quickly login like seen here:

login

### CODE REVIEW

File: Source/UI/IdentityUI.swift [hunk 1]
```diff
@@ -753,3 +774,45 @@ private final class DismissableNavigationController: UINavigationController {
...
 }
+
+private final class SharedWebCredentialsController: NSObject {
+    // Stored anonymously to retain the ASAuthorizationController
+    private var authorizationController: NSObject?
+
+    var didCompleteWithAuthorization: (String, String) -> Void = { _, _ in }
+    var didCompleteWithError: (Swift.Error) -> Void = { _ in }
+
+    override init() {
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:796]: Can we provide the result as a `Swift.Result` and add closure to the initialiser?

File: Source/UI/IdentityUI.swift [hunk 2]
```diff
@@ -462,6 +468,21 @@ extension IdentityUI {
...
+                sharedWebCredentialsController.didCompleteWithError = { [weak self] _ in
+                    self?.show(input: .byLoginMethod(.password, presentingViewController: vc, localizedTeaserText: localizedTeaserText, scopes: scopes), client: client, completion: completion)
+                }
+                sharedWebCredentialsController.didCompleteWithAuthorization = { [weak self] user, password in
+                    guard let email = EmailAddress(user) else { return }
+                    self?._identityManager.login(username: .email(email), password: password, scopes: scopes, persistUser: true, useSharedWebCredentials: false, completion: { _ in
+                        completion(.skip)
+                    })
+                }
+                self.sharedWebCredentialsController = sharedWebCredentialsController
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:493]: Note: `authorizationController` has to be reset to avoid the retain cycle.

File: Source/UI/IdentityUI.swift [hunk 3]
```diff
@@ -462,6 +468,21 @@ extension IdentityUI {
         case let .byRoute(route, vc):
             handleRouteForUnpresentedUI(route: route, byPresentingIn: vc, client: client, completion: completion)
         case let .byLoginMethod(loginMethod, vc, localizedTeaserText, scopes):
+            if #available(iOS 13.0, *), case .sharedWebCredentials = loginMethod {
+                let sharedWebCredentialsController = SharedWebCredentialsController()
+                sharedWebCredentialsController.didCompleteWithError = { [weak self] _ in
+                    self?.show(input: .byLoginMethod(.password, presentingViewController: vc, localizedTeaserText: localizedTeaserText, scopes: scopes), client: client, completion: completion)
+                }
+                sharedWebCredentialsController.didCompleteWithAuthorization = { [weak self] user, password in
+                    guard let email = EmailAddress(user) else { return }
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift]: Here we have to fall back as well

File: Source/UI/IdentityUI.swift [hunk 4]
```diff
@@ -136,6 +140,8 @@ public class IdentityUI {
     }
     private let _identityManager: IdentityManager
 
+    private var sharedWebCredentialsController: SharedWebCredentialsController?
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:143]: Can we avoid storing it? It should exist during fetching time only

File: Source/UI/IdentityUI.swift (hunk 2 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:493]: What does "reset" mean in this context? You mean we have to explicitly set it to nil?

File: Source/UI/IdentityUI.swift (hunk 4 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:143]: We need to retain it as the delegate is weak.

Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:143]: But I've changed it to be removed (nilled out) once the completion blocks are done.

File: Source/UI/IdentityUI.swift [hunk 5]
```diff
@@ -753,3 +789,45 @@ private final class DismissableNavigationController: UINavigationController {
         fatalError("init(coder:) has not been implemented")
     }
 }
+
+private final class SharedWebCredentialsController: NSObject {
+    // Stored anonymously to retain the ASAuthorizationController
+    private var authorizationController: NSObject?
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:795]: Let's just do `let controller = ASAuthorizationController(authorizationRequests: [ASAuthorizationPasswordProvider().createRequest()])` and don't bother clearing it out as it will be gone along with `SharedWebCredentialsController`

File: Source/UI/IdentityUI.swift [hunk 6]
```diff
@@ -462,6 +468,36 @@ extension IdentityUI {
         case let .byRoute(route, vc):
             handleRouteForUnpresentedUI(route: route, byPresentingIn: vc, client: client, completion: completion)
         case let .byLoginMethod(loginMethod, vc, localizedTeaserText, scopes):
+            if #available(iOS 13.0, *), case .sharedWebCredentials = loginMethod {
+                let fallback = { [weak self] in
+                    self?.show(input: .byLoginMethod(.password, presentingViewController: vc, localizedTeaserText: localizedTeaserText, scopes: scopes), client: client, completion: completion)
+                    self?.sharedWebCredentialsController = nil
+                }
+
+                sharedWebCredentialsController = SharedWebCredentialsController { [weak self] result in
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:477]: nit: we can have `sharedWebCredentialsController` captured by the closure, to it leaves until the closure it executed, but then we would have to provide a closure separately, not via the initialiser

File: Source/UI/IdentityUI.swift (hunk 5 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:795]: We can't have a strong typed class member since it would require the entire class to be marked `@available(iOS 13.0, *)` which then cascades up. This is intentionally kept as anonymous type.

File: Source/UI/IdentityUI.swift (hunk 6 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:477]: Yeah, but I dislike the idea of _intentionally_ using a closure to retain an object, rather than having a temporary class member.

File: Source/UI/IdentityUI.swift [hunk 7]
```diff
@@ -23,6 +24,9 @@ public enum LoginMethod {
     /// asks for identifier and then a password to either login or signup if not already registered; the specified email will appear pre-filled in the identity
     /// UI, yet the user will still be able to modify it before submitting.
     case passwordWithPrefilledEmail(EmailAddress)
+    /// prompts the user for using their shared web credentials (if available)
+    @available(iOS 13.0, *)
+    case sharedWebCredentials
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:29]: Shouldn't these be called just `sharedCredentials` -- why are they `web`?

File: Source/UI/IdentityUI.swift [hunk 8]
```diff
@@ -200,7 +206,7 @@ public class IdentityUI {
         let input = UpdatedTermsCoordinator.Input(currentUser: user, terms: terms)
 
         updatedTermsCoordinator = UpdatedTermsCoordinator(navigationController: navigationController, configuration: configuration)
-        updatedTermsCoordinator?.start(input: input) { _ in
+        updatedTermsCoordinator?.start(input: input) { output in
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:209]: nit: where the newly introduced `output` is used?

File: Source/UI/IdentityUI.swift [hunk 9]
```diff
@@ -462,6 +468,11 @@ extension IdentityUI {
         case let .byRoute(route, vc):
             handleRouteForUnpresentedUI(route: route, byPresentingIn: vc, client: client, completion: completion)
         case let .byLoginMethod(loginMethod, vc, localizedTeaserText, scopes):
+            if #available(iOS 13.0, *), case .sharedWebCredentials = loginMethod {
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:471]: nit: Why do we have to additionally check available version as `sharedWebCredentials` is available as of iOS 13?

File: Source/UI/IdentityUI.swift [hunk 10]
```diff
@@ -477,6 +488,105 @@ extension IdentityUI {
...
+                 completion: completion)
+
+            sharedWebCredentialsController = nil
+        }
+
+        sharedWebCredentialsController = SharedWebCredentialsController { [weak self] result in
+            guard case .success((let username, let password)) = result,
+                  let email = EmailAddress(username),
+                  let self = self else {
+                fallback()
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:512]: nit: Can we log the error here? It would be good to know whether it's just because of cancellation or invalid user name (malformed e-mail)

File: Source/UI/IdentityUI.swift [hunk 11]
```diff
@@ -477,6 +488,105 @@ extension IdentityUI {
...
+                }
+
+                func cancel() {
+                    completion(.cancel)
+                }
+
+                func presentTerms() {
+                    self.identityManager.fetchTerms { result in
+                        guard case .success(let terms) = result else {
+                            cancel()
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:539]: Does it mean we will be just silently fail if we fail to fetch terms? Shall we logout in this case as well?

File: Source/UI/IdentityUI.swift [hunk 12]
```diff
@@ -477,6 +488,105 @@ extension IdentityUI {
...
+
+                        updatedTermsCoordinator.start(input: .init(currentUser: user, terms: terms)) { output in
+                            IdentityUI.updatedTermsCoordinator = nil
+
+                            navigationController.dismiss(animated: true) {
+                                switch output {
+                                case .success:
+                                    success()
+                                case .cancel:
+                                    cancel()
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:565]: Same thing w/ logout

File: Source/UI/IdentityUI.swift [hunk 13]
```diff
@@ -477,6 +488,105 @@ extension IdentityUI {
...
+                    }
+                }
+
+                user.agreements.status { result in
+                    guard case .success(let agreed) = result, agreed else {
+                        presentTerms()
+                        return
+                    }
+
+                    success()
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:582]: Why do we success if agreements cannot be checked?

File: Source/UI/Screens/Identifier/IdentifierViewController.swift [hunk 14]
```diff
@@ -170,6 +170,8 @@ class IdentifierViewController: IdentityUIViewController {
             showPhoneNumber()
             countryCode.text = prefilledPhoneComponents.countryCode
             phoneNumber.text = prefilledPhoneComponents.number
+        case .sharedWebCredentials:
+            fatalError()
```
Comment (vmagaziy) [inline Source/UI/Screens/Identifier/IdentifierViewController.swift:174]: What does this stand for?

File: Source/UI/IdentityUI.swift [hunk 15]
```diff
@@ -477,6 +488,105 @@ extension IdentityUI {
...
+                return
+            }
+
+            self.identityManager.login(username: .email(email),
+                                       password: password,
+                                       scopes: scopes,
+                                       persistUser: true,
+                                       useSharedWebCredentials: false) { result in
+                guard case .success = result else {
+                    fatalError()
```
Comment (vmagaziy) [inline Source/UI/IdentityUI.swift:522]: Where is the guarantee login is always succeeding? What if user with stored e-mail is no longer valid or password got changed?

File: Source/UI/IdentityUI.swift (hunk 7 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:29]: No, same naming convention as #314

File: Source/UI/IdentityUI.swift (hunk 9 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:471]: Because you can't call a iOS 13 method without a #available check. It's required by the compiler.

File: Source/UI/IdentityUI.swift (hunk 8 above)
Comment (clausjoergensen) [inline Source/UI/IdentityUI.swift:209]:
```suggestion
        updatedTermsCoordinator?.start(input: input) { _ in
```

File: Source/UI/IdentityUI.swift (hunk 13 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:582]: We're not, you're reading the code wrong. We succeed if we read the agreements _and_ they were agreed to.

File: Source/UI/IdentityUI.swift (hunk 12 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:565]: The logout is already called inside the `UpdatedTermsCoordinator`

File: Source/UI/IdentityUI.swift (hunk 11 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:539]: The existing flow doesn't logout in this case, but it does present a popup with the error. Maybe we should do the same here.

File: Source/UI/IdentityUI.swift (hunk 15 above)
Comment (clausjoergensen) [inline Source/UI/IdentityUI.swift:522]:
```suggestion
                    fallback()
```

Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift]: This was a typo, meant to be `fallback()`

File: Source/UI/Screens/Identifier/IdentifierViewController.swift (hunk 14 above)
Comment (clausjoergensen) [reply Source/UI/Screens/Identifier/IdentifierViewController.swift:174]: This flow is impossible to hit (the view controller will never be presented with this login method)

**Comment by vmagaziy** [comment]
Something is wrong with styling when T&C are presented in the flow:

Screen Shot 2020-11-18 at 10 13 26

**Comment by clausjoergensen** [comment]
> Something is wrong with styling when T&C are presented in the flow:

Everything looks correct to me on that screenshot.

**Comment by vmagaziy** [comment]
The properly styled T&C page looks like this:

Screen Shot 2020-11-18 at 10 19 47

Notice the colours of buttons and checkboxes.

**Comment by clausjoergensen** [comment]
@vmagaziy the theme have been fixed now.

File: Source/UI/IdentityUI.swift (hunk 11 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift]: This now logs out the user and presents a popup with an error.

File: Source/Core/SynchronizedWeakDictionary.swift [hunk 16]
```diff
@@ -18,19 +18,19 @@ class WeakValue<V: AnyObject>: CustomStringConvertible {
...
+            return weakValue.value
+        }
         set {
             guard newValue != nil else {
                 dictionary[key] = nil
                 return
             }
             dictionary[key] = WeakValue(newValue)
         }
-        get {
```
Comment (vmagaziy) [inline Source/Core/SynchronizedWeakDictionary.swift:28]: Here

File: Source/Manager/IdentityManager.swift [hunk 17]
```diff
@@ -370,7 +370,8 @@ public class IdentityManager: IdentityManagerProtocol {
      - parameter useSharedWebCredentials: whether the credentials should be saved to the shared web credentials.
      - parameter completion: a callback that is called after the credential is checked.
      */
-    public func login(username: Identifier, password: String, scopes: [String] = [], persistUser: Bool, useSharedWebCredentials: Bool, completion: @escaping NoValueCallback) {
+    public func login(username: Identifier, password: String, scopes: [String] = [], persistUser: Bool,
```
Comment (vmagaziy) [inline Source/Manager/IdentityManager.swift:373]: Same

File: Source/Manager/IdentityManagerOverloads.swift [hunk 18]
```diff
@@ -40,8 +40,10 @@ public extension IdentityManagerProtocol {
     }
 
     /// - SeeAlso: `IdentityManager.login(...)`
-    func login(email: EmailAddress, password: String, scopes: [String] = [], persistUser: Bool, useSharedWebCredentials: Bool, completion: @escaping NoValueCallback) {
-        return login(username: Identifier(email), password: password, scopes: scopes, persistUser: persistUser, useSharedWebCredentials: useSharedWebCredentials, completion: completion)
+    func login(email: EmailAddress, password: String, scopes: [String] = [], persistUser: Bool,
```
Comment (vmagaziy) [inline Source/Manager/IdentityManagerOverloads.swift:43]: Same

File: Source/Manager/LegacySPiDAccessToken/SPiDAccessToken.swift [hunk 19]
```diff
@@ -50,7 +50,7 @@ open class SPiDAccessToken: NSObject, NSCoding {
         coder.encode(expiresAt, forKey: CodingKeys.expiresAt.rawValue)
         coder.encode(refreshToken, forKey: CodingKeys.refreshToken.rawValue)
     }
-
```
Comment (vmagaziy) [inline Source/Manager/LegacySPiDAccessToken/SPiDAccessToken.swift:53]: Same

File: Source/Manager/Protocols/IdentityManagerProtocol.swift [hunk 20]
```diff
@@ -26,7 +26,8 @@ public protocol IdentityManagerProtocol: AnyObject {
     func validate(oneTimeCode: String, for _: Identifier, scopes: [String], persistUser: Bool, completion: @escaping NoValueCallback)
 
     ///
-    func login(username: Identifier, password: String, scopes: [String], persistUser: Bool, useSharedWebCredentials: Bool, completion: @escaping NoValueCallback)
+    func login(username: Identifier, password: String, scopes: [String], persistUser: Bool,
```
Comment (vmagaziy) [inline Source/Manager/Protocols/IdentityManagerProtocol.swift:29]: Same

File: Source/UI/Coordinators/PasswordCoordinator.swift [hunk 21]
```diff
@@ -104,7 +104,13 @@ extension PasswordCoordinator {
         viewController.didRequestAction = { [weak self] action in
             switch action {
             case let .enter(password, shouldPersistUser):
-                self?.submit(password: password, for: identifier, on: loginFlowVariant, persistUser: shouldPersistUser, scopes: scopes, useSharedWebCredentials: useSharedWebCredentials, completion: completion)
+                self?.submit(password: password,
```
Comment (vmagaziy) [inline Source/UI/Coordinators/PasswordCoordinator.swift:107]: Same

File: Source/UI/Coordinators/PasswordCoordinator.swift [hunk 22]
```diff
@@ -148,7 +154,8 @@ extension PasswordCoordinator {
         }
 
         presentedViewController?.startLoading()
-        signinInteractor.login(username: identifier, password: password, scopes: scopes, useSharedWebCredentials: useSharedWebCredentials) { [weak self] result in
+        signinInteractor.login(username: identifier, password: password, scopes: scopes,
```
Comment (vmagaziy) [inline Source/UI/Coordinators/PasswordCoordinator.swift:157]: Same

**Review by vmagaziy — COMMENTED** [review]
@clausjoergensen -- can we reduce a number of changes, so we have only 2 files affected as before + required adjustments to tests?

**Comment by clausjoergensen** [comment]
> @clausjoergensen -- can we reduce a number of changes, so we have only 2 files affected as before + required adjustments to tests?

No. We need the project to build to ensure full testing (and I'm sure you can review a few whitespace changes)

**Comment by clausjoergensen** [comment]
I moved all the cleanup to #322

**Review by vmagaziy — APPROVED** [review]
lgtm

File: Source/UI/IdentityUI.swift [hunk 23]
```diff
@@ -477,6 +491,115 @@ extension IdentityUI {
...
+                                            scopes: scopes,
+                                            persistUser: true,
+                                            useSharedWebCredentials: false) { result in
+                    switch result {
+                    case .failure(let error):
+                        log(from: self, error)
+                        fallback(.passwordWithPrefilledEmail(email))
+                    case .success:
+                        let user = self._identityManager.currentUser
+                        user.agreements.status { result in
```
Comment (zamzterz) [inline Source/UI/IdentityUI.swift:584]: Instead of this re-implementation of asking user to accept agreements, I think `CompleteProfileCoordinator` could be re-used.

Something like:
```swift
let coordinator = CompleteProfileCoordinator(navigationController: navigationController,
                                             identityManager: self._identityManager,
                                             configuration: self.configuration)
let profileCoordinator = UpdateProfileInteractor(currentUser: user,
                                                 loginFlowVariant: .signin,
                                                 tracker: self.configuration.tracker)
```

**Review by zamzterz — CHANGES_REQUESTED** [review]
Nice one! 👍 

Unfortunately there's more complexity to handle in the case of successful login:
This correctly handles accepting terms, but there might also be required fields the user needs to enter data for. This is pretty much what `CompleteProfileCoordinator` captures. So I think that needs to be triggered/displayed in case of successful login with password from shared web credentials.

File: Source/UI/IdentityUI.swift (hunk 23 above)
Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:584]: The problem is we can't present the other coordinators from this flow, as they don't have a completion block. This results in the consumer of the API (the app) doesn't know whether the login succeeded or not, and can't update it's UI.

Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:584]: And remember, we also want to do this without showing any UI unless necessary. From what I've seen, the coordinators flow _always_ presents something.

Comment (clausjoergensen) [reply Source/UI/IdentityUI.swift:584]: Ah, CompleteProfileCoordinator is special in that it checks all the paths and doesn't present anything unless it needs to. (That's different from the UpdateProfileCoordinator). That might work.

**Comment by clausjoergensen** [comment]
@zamzterz I pushed a update using CompleteProfileCoordinator, I've verified it with the Terms&Conditions flow, seems to work.

**Comment by clausjoergensen** [comment]
Hmm, it's not resetting correctly, so subsequent login attempts are failing identityUICompletion, looks like using the DismissableNavigationController doesn't work with the CompleteProfileCoordinator

**Comment by clausjoergensen** [comment]
@zamzterz ah, no. Using this coordinator breaks the login flow because it attempts to present the UI. We can't have it present _anything_, no spinner, no nothing, unless it has to. 

This is what it looks like in a successful scenario for a user who don't need to agree / accept anything:

IMG_0591

**Review by zamzterz — APPROVED** [review]
